src/stream/src/executor/project_set.rs

```diff
@@ -97,107 +96,93 @@ impl Executor for ProjectSetExecutor {
-                            })
-                            .collect_vec();
-
-                        // The maximum length of the results of table functions will be the output
-                        // length.
-                        let max_tf_len = items
-                            .iter()
-                            .map(|i| i.as_ref().map_left(|arr| arr.len()).left_or(0))
-                            .max()
-                            .unwrap();
-
+                    // for each input row
+                    for row_idx in 0..chunk.capacity() {
                         // ProjectSet cannot preserve that U- is followed by U+,
                         // so we rewrite update to insert/delete.
-                        let op = match op {
+                        let op = match chunk.ops()[row_idx] {
                             Op::Delete | Op::UpdateDelete => Op::Delete,
                             Op::Insert | Op::UpdateInsert => Op::Insert,
                         };
-                        ret_ops.extend(vec![op; max_tf_len]);
-                        for i in 0..max_tf_len {
-                            projected_row_id_builder.append(Some(i as i64));
-                        }
-
-                        for (item, builder) in items.into_iter().zip_eq_fast(builders.iter_mut()) {
-                            match item {
-                                Either::Left(array_ref) => {
-                                    builder.append_array(&array_ref);
-                                    for _ in 0..(max_tf_len - array_ref.len()) {
-                                        builder.append_null();
-                                    }
-                                }
-                                Either::Right(datum_ref) => {
-                                    for _ in 0..max_tf_len {
-                                        builder.append_datum(datum_ref);
+                        // for each output row
+                        for projected_row_id in 0i64.. {
+                            // SAFETY:
+                            // We use `row` as a buffer and don't read elements from the previous
+                            // loop. The `transmute` is used for bypassing the borrow checker.
+                            let row: &mut [DatumRef<'_>] =
+                                unsafe { std::mem::transmute(row.as_mut_slice()) };
+
+                            row[0] = Some(projected_row_id.into());
+                            // if any of the set columns has a value
+                            let mut valid = false;
+                            // for each column
+                            for (item, value) in results.iter_mut().zip_eq_fast(&mut row[1..]) {
+                                *value = match item {
+                                    Either::Left(state) => if let Some((i, value)) = state.peek() && i == row_idx {
+                                        valid = true;
```

Comment:
Can we directly call `next` here?

Reply:
Nope. It's a bit tricky here. Since we want to get a `DatumRef` which borrows the data chunk, it can not be made as standard `Iterator` or `Stream`. The state can not move to the next row until we release the `DatumRef`.